store/multi_writer_app_store.go

```diff
@@ -97,6 +98,80 @@ func NewMultiWriterAppStore(appStore *IAVLStore, evmStore *EvmStore, saveEVMStat
 	return store, nil
 }
 
+func LoadMultiWriterAppStore(appstoreCfg *AppStoreConfig, evmstoreCfg *EvmStoreConfig, appstoredb db.DBWrapper,
+	targetVersion int64, metricsDatabase bool, rootPath string) (*MultiWriterAppStore, error) {
+	iavlStore, err := NewIAVLStore(appstoredb, appstoreCfg.MaxVersions, targetVersion, appstoreCfg.IAVLFlushInterval)
+	if err != nil {
+		return nil, err
+	}
+	appStoreEvmRoot := iavlStore.Get(rootKey)
+	evmDB, err := db.LoadDB(
+		evmstoreCfg.DBBackend,
+		evmstoreCfg.DBName,
+		rootPath,
+		evmstoreCfg.CacheSizeMegs,
+		evmstoreCfg.WriteBufferMegs,
+		metricsDatabase,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	evmStore, err := loadEvmStore(evmDB, iavlStore.Version(), evmstoreCfg.NumCachedRoots)
+	if err != nil {
+		return nil, err
+	}
+	evmStoreEvmRoot, evmVersion := evmStore.getLastSavedRoot(iavlStore.Version())
+	if !bytes.Equal(appStoreEvmRoot, evmStoreEvmRoot) {
+		log.Info(
+			"EVM root mismatch, resyncing roots...",
+			"evm.db-ver", evmVersion, "app.db-ver", iavlStore.Version(),
+			"evm.db-root", hex.EncodeToString(evmStoreEvmRoot), "app.db-evm-root", hex.EncodeToString(appStoreEvmRoot),
+		)
+		latestVersion, err := iavl.NewMutableTree(appstoredb, 10000).LatestVersion(evmVersion)
+		if err != nil {
+			return nil, err
+		}
+		if latestVersion == 0 {
+			latestVersion = -1
+		}
+		iavlStore, err = NewIAVLStore(appstoredb, appstoreCfg.MaxVersions, latestVersion, appstoreCfg.IAVLFlushInterval)
+		if err != nil {
+			return nil, err
+		}
+		evmStore.CloseDB()
+		evmDB, err = db.LoadDB(
+			evmstoreCfg.DBBackend,
+			evmstoreCfg.DBName,
+			rootPath,
+			evmstoreCfg.CacheSizeMegs,
+			evmstoreCfg.WriteBufferMegs,
+			metricsDatabase,
+		)
+		if err != nil {
+			return nil, err
+		}
+		evmStore, err = loadEvmStore(evmDB, iavlStore.Version(), evmstoreCfg.NumCachedRoots)
+		if err != nil {
+			return nil, err
+		}
+	}
+	store := &MultiWriterAppStore{
```

Comment:
We should check EVM root one more time to make sure that EVM roots are the same. So instead of copy the logic from `NewMultiWriterAppStore`, we can do
```
return NewMultiWriterAppStore(iavlStore, evmStore, appstoreCfg.SaveEVMStateToIAVL)
```
so we can delete the logic below this line.